tacs/problems/static.py

```diff
@@ -754,86 +896,217 @@ def solve(self, Fext=None):
+                self._pp("|")
+                self._pp(
+                    "| %-30s: %10.3f sec"
+                    % ("TACS Total Solution Time", finalNormTime - startTime)
+                )
+                self._pp("+--------------------------------------------------+")
+
+        return hasConverged
+
+    def solveNonlinear(self, Fext=None):
+        # Compute the internal and external force components of the residual at the current point
+        self.getForces(
+            externalForceVec=self.externalForce,
+            internalForceVec=self.internalForce,
+            Fext=Fext,
+        )
+        self.initNorm = np.real(self.externalForce.norm())
+
+        # We need to update the residual function handle used by the nonlinear solver based on the current external force vector
+        def resFunc(res):
+            self.getResidual(res, Fext=Fext)
+
+        self.continuationSolver.resFunc = resFunc
+        self.newtonSolver.resFunc = resFunc
+        self.continuationSolver.setRefNorm(self.initNorm)
+        self.continuationSolver.solve()
+
+        # Since the state has changed, we need to flag that the jacobian and preconditioner should be before the next primal or adjoint solve
+        self._preconditionerUpdateRequired = True
+        self._jacobianUpdateRequired = True
+
+        # We should also reset the linear solver convergence so that any loosening of the linear convergence criteria are undone before a possible adjoint solve
+        self.KSM.setTolerances(
+            self.getOption("L2ConvergenceRel"),
+            self.getOption("L2Convergence"),
+        )
+
+        # Finally return a bool indicating whether the solve was successful
+        return self.continuationSolver.hasConverged
+
+    def _nonlinearCallback(self, solver, u, res, monitorVars):
+        """Callback function to be called by the nonlinear solver at each iteration
+
+        Parameters
+        ----------
+        solver : pyTACS solver object
+            The solver
+        u : tacs.TACS.Vec
+            Current state vector
+        res : tacs.TACS.Vec
+            Current residual vector
+        monitorVars : dict
+            Dictionary of variables to monitor, the values the solver should include can be
+            specified through the ``"nonlinearSolverMonitorVars"`` option.
+        """
+        iteration = 0
+        if self.rank == 0:
+            # Figure out the iteration number
+            iteration = self.history.getIter()
+            if iteration % 50 == 0:
+                self.history.printHeader()
+            self.history.write(monitorVars)
+            self.history.printData()
+
+        self.comm.bcast(iteration, root=0)
+
+        if self.getOption("writeNLIterSolutions"):
+            # import pdb
+
+            # pdb.set_trace()
+            self.writeSolution(
+                baseName=f"{self.name}-{self.callCounter:03d}-NLIter", number=iteration
+            )
+
+    def _solveLinear(self, res, sol):
```

Comment:
Not sure this it's possible to make the API's the same because the two methods do fundamentally different things. The linear equivalent of `solveNonlinear` would probably do all of [this](https://github.com/smdogroup/tacs/blob/1d54bd3270d3c518c0019978928482accc814cf3/tacs/problems/static.py#L902-L929). I could make a `_solveLinear` that does all that but I'd like there to still be a function that does what the current `_solveLinear` does, just solving the linear system with an arbitrary right hand side. Any suggestions on what to call that?

Reply:
How about `_solveLinearProblem` and `solveNonlinearProblem` to indicate that they do some things that are specific to this problem type, while `_solveLinear` is more generic?

Reply:
I think that will be assumed since this is a method of the problem class. Having "problem" in the method name seems redundant to me

Reply:
Addressed in https://github.com/smdogroup/tacs/pull/193/commits/c5c297464ec1ac883845ece9a036d24d1b53c234